docx/footnotes.py

```diff
@@ -0,0 +1,84 @@
+# encoding: utf-8
+
+"""
+Footnotes-related proxy types.
+"""
+
+from .shared import Parented
+from .blkcntnr import BlockItemContainer
+
+
+class Footnotes(Parented):
+    """
+    Proxy object wrapping ``<w:footnotes>`` element.
+    """
+    def __init__(self, footnotes, parent):
+        super(Footnotes, self).__init__(parent)
+        self._element = self._footnotes = footnotes
+
+    def __getitem__(self, reference_id):
+        """
+        A |Footnote| for a specific footnote of reference id, defined with ``w:id`` argument of ``<w:footnoteReference>``.
+        If reference id is invalid raises an |IndexError|
+        """
+        footnote = self._element.get_by_id(reference_id)
+        if footnote is None:
+            raise IndexError
+        return Footnote(footnote, self)
+
+    def __len__(self):
+        return len(self._element)
+
+    def add_footnote(self, footnote_reference_id):
+        """
+        Return a newly created |Footnote|, the new footnote will
+        be inserted in the correct spot by `footnote_reference_id`.
+        The footnotes are kept in order by `footnote_reference_id`.
+        """
+        elements = self._element # for easy access
+        new_footnote = None
+        if elements.get_by_id(footnote_reference_id):
+            # When adding a footnote it can be inserted 
+            # in front of some other footnotes, so
+            # we need to sort footnotes by `footnote_reference_id`
+            # in |Footnotes| and in |Paragraph|
+            #
+            # resolve reference ids in |Footnotes|
+            # iterate in reverse and compare the current
+            # id with the inserted id. If there are the same
+            # insert the new footnote in that place, if not
+            # increment the current footnote id.
```

Comment:
This micro-optimization makes sense, but is depends on case-by-case basis. If user only adds footnotes at the beginning, we don't achieve any performance improvement. I'm not saying we should change it, just leaving a comment so we remember to take care of this in the future. We need some benchmarks and usage scenarios to conclude the best scenario.

Reply:
I agree, the optimization is based on premise that the draft server will add footnotes on the end of the document.